adapters/improvedigital/improvedigital.go

```diff
@@ -144,3 +147,77 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
 		Message: fmt.Sprintf("Failed to find impression for ID: \"%s\"", impID),
 	}
 }
+
+// This method responsible to clone request and convert additional consent providers string to array when additional consent provider found
+func (a *ImprovedigitalAdapter) applyAdditionalConsentProvidersArray(request openrtb2.BidRequest) openrtb2.BidRequest {
+	// If user/user.ext not defined, no need to parse additional consent
+	if request.User == nil || request.User.Ext == nil {
+		return request
+	}
+
+	var (
+		err     error
+		reqJSON json.RawMessage
+		cpStr   string
+	)
+
+	// Start validating additional consent
+	// Check key exist user.ext.ConsentedProvidersSettings.consented_providers
+	var userExtMap = make(map[string]json.RawMessage)
+	if err = json.Unmarshal(request.User.Ext, &userExtMap); err != nil {
+		return request
+	}
+
+	cpsMapValue, cpsJSONFound := userExtMap["ConsentedProvidersSettings"]
+	if !cpsJSONFound {
+		return request
+	}
+
+	// Check key exist user.ext.ConsentedProvidersSettings.consented_providers
+	var cpMap = make(map[string]json.RawMessage)
+	if err = json.Unmarshal(cpsMapValue, &cpMap); err != nil {
+		return request
+	}
+
+	cpMapValue, cpJSONFound := cpMap["consented_providers"]
+	if !cpJSONFound {
+		return request
+	}
+	// End validating additional consent
+
+	// Check if string contain ~, then substring after ~ to end of string
+	consentStr := string(cpMapValue)
+	var tildaPosition int
+	if tildaPosition = strings.Index(consentStr, "~"); tildaPosition == -1 {
+		return request
+	}
+	cpStr = consentStr[tildaPosition+1 : len(consentStr)-1]
+
+	// Prepare consent providers string for append
+	cpStr = fmt.Sprintf("[%s]", strings.Replace(cpStr, ".", ",", -1))
+	cpMap["consented_providers"] = json.RawMessage(cpStr)
+
+	cpJSON, _ := json.Marshal(cpMap)
+	userExtMap["consented_providers_settings"] = cpJSON
+
+	extJson, extErr := json.Marshal(userExtMap)
+	if extErr != nil {
+		return request
+	}
+
+	// Start request closing
+	reqJSON, err = json.Marshal(request)
+	if err != nil {
+		return request
+	}
+
+	var reqClone openrtb2.BidRequest
+	if err = json.Unmarshal(reqJSON, &reqClone); err != nil {
+		return request
+	}
+	// End request closing
+
+	reqClone.User.Ext = extJson
```

Comment:
Instead of Unmarshal request to clone request and then Marshal it back right after this function call, would it be easier to return updated User ext and set it if it was changed after `cpStr` evaluation?

Reply:
It would be better if you return error and data (request or user.ext part of it).

Reply:
We can't change user.ext of original request (test will fail). So, we must have to masshal/unmarshal it again.

Sure, we will update our code so it will return data and error (request and error)

Reply:
Right, you need to make request to have it's own User copy:
```
newUserExt := a.applyAdditionalConsentProvidersArray(request) <- returns []byte
	if len(newUserExt) > 0 {
		userCopy := *request.User
		userCopy.Ext = newUserExt
		request.User = &userCopy
	}
```
You will not need to create a clone of the entire request.
Here is a link to documentations: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#makerequests , navigate to `Example: Mutating banner shared memory (make a copy).`